lib/matplotlib/_api/deprecation.pyi

```diff
@@ -0,0 +1,70 @@
+from collections.abc import Callable
+import contextlib
+from typing import Literal, TypedDict, TypeVar, overload
+from typing_extensions import (
+    ParamSpec,  # < Py 3.10
+    Unpack,  # < Py 3.11
+)
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+class MatplotlibDeprecationWarning(DeprecationWarning): ...
+
+class DeprecationKwargs(TypedDict, total=False):
+    message: str
+    alternative: str
+    pending: bool
+    obj_type: str
+    addendum: str
+    removal: str
+
+class NamedDeprecationKwargs(DeprecationKwargs, total=False):
+    name: str
+
+def warn_deprecated(since: str, **kwargs: Unpack[NamedDeprecationKwargs]) -> None: ...
+def deprecated(since: str, **kwargs: Unpack[NamedDeprecationKwargs]) -> Callable: ...
+
+class deprecate_privatize_attribute:
+    def __init__(self, since: str, **kwargs: Unpack[NamedDeprecationKwargs]): ...
+    def __set_name__(self, owner: type[object], name: str) -> None: ...
+
+DECORATORS: dict[Callable, Callable] = ...
+
+@overload
+def rename_parameter(
+    since: str, old: str, new: str, func: Literal[None] = None
+) -> Callable[[Callable[P, R]], Callable[P, R]]: ...
+@overload
+def rename_parameter(func: Callable[P, R]) -> Callable[P, R]: ...
```

Comment:
I think this needs all of `since`, `old` and `new` still...

There is no call pattern for `rename_parameter` that does not take these arguments, there is one that directly  returns the decorated function handed as a parameter or one that returns the inner decorator.

Additionally, not actually sure that the paramspec here is correct (or even _representable_):

[PEP 612](https://peps.python.org/pep-0612/#concatenating-keyword-parameters) explicitly does _not_ support appending keyword-only arguments (it allows prepending/appending positional args)

Again, not actually too upset if a deprecation only type hints as one of the paths (preferably the new path when its not just going away completely), so not the worst.

So, I guess maybe this is the best we can actually do...

I may suggest making the typevars private, though

Reply:
The input function to the `rename_parameter` decorator should be using the new name already. I didn't want to add the old name to the types either.

Reply:
Re: the overload; I thought something didn't seem to like it, but looking at it again, I'm not sure I even need the second one. The first signature with the `Callable` in/out seems sufficient.

Reply:
> Re: the overload; I thought something didn't seem to like it, but looking at it again, I'm not sure I even need the second one. The first signature with the `Callable` in/out seems sufficient.

Oh, scratch that latter bit, I of course messed up the type as `Literal[None]`; should be fixed now.